File: iot/amqp/sections/AMQPProperties.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.avps.SectionCode as SectionCode
import iot.amqp.wrappers.LongID as LongID
import iot.amqp.wrappers.StringID as StringID
import iot.amqp.wrappers.BinaryID as BinaryID
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVList as TLVList

class amqpProperties():
    def __init__(self,messageId,userId,to,subject,replyTo,correlationId,contentType,contentEncoding,absoluteExpiryTime,creationTime,groupId,groupSequence,replyToGroupId):
        self.messageId = messageId
        self.userId = userId
        self.to = to
        self.subject = subject
        self.replyTo = replyTo
        self.correlationId = correlationId
        self.contentType = contentType
        self.contentEncoding = contentEncoding
        self.absoluteExpiryTime = absoluteExpiryTime
        self.creationTime = creationTime
        self.groupId = groupId
        self.groupSequence = groupSequence
        self.replyToGroupId = replyToGroupId

    def getValue(self):
        list  = TLVList.tlvList(None, None)
        wrapper = AMQPWrapper.amqpWrapper()

        if self.messageId is not None:
            value = None
            if self.messageId.getBinary() is not None:
                value = self.messageId.getBinary()
            elif self.messageId.getLong() is not None:
                value = self.messageId.getLong()
            elif self.messageId.getString() is not None:
                value = self.messageId.getString()
            elif self.messageId.getUuid() is not None:
                value = self.messageId.getUuid()
            list.addElement(0, wrapper.wrap(value))

        if self.userId is not None:
            list.addElement(1, wrapper.wrap(self.userId))
        if self.to is not None:
            list.addElement(2, wrapper.wrap(self.to))
        if self.subject is not None:
            list.addElement(3, wrapper.wrap(self.subject))
        if self.replyTo is not None:
            list.addElement(4, wrapper.wrap(self.replyTo))
        if self.correlationId is not None:
            list.addElement(5, wrapper.wrap(self.correlationId))
        if self.contentType is not None:
            list.addElement(6, wrapper.wrap(self.contentType))
        if self.contentEncoding is not None:
            list.addElement(7, wrapper.wrap(self.contentEncoding))
        if self.absoluteExpiryTime is not None:
            list.addElement(8, wrapper.wrap(self.absoluteExpiryTime))
        if self.creationTime is not None:
            list.addElement(9, wrapper.wrap(self.creationTime))
        if self.groupId is not None:
            list.addElement(10, wrapper.wrap(self.groupId))
        if self.groupSequence is not None:
            list.addElement(11, wrapper.wrap(self.groupSequence))
        if self.replyToGroupId is not None:
            list.addElement(12, wrapper.wrap(self.replyToGroupId))

        constructor = DescribedConstructor.describedConstructor(list.getCode(), TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), 0x73))
        list.setConstructor(constructor)
        return list

    def fill(self, value):
        list = value
        unwrapper = AMQPUnwrapper.amqpUnwrapper()

        if len(list.getList()) > 0:
            element = list.getList()[0]
            if element is not None:
                code = element.getCode()
                if code in (AMQPType.amqpType.getValueByKey('ULONG_0'),AMQPType.amqpType.getValueByKey('SMALL_ULONG'),AMQPType.amqpType.getValueByKey('ULONG')):
                    self.messageId = LongID.longID(unwrapper.unwrapULong(element))
                elif code in (AMQPType.amqpType.getValueByKey('STRING_8'),AMQPType.amqpType.getValueByKey('STRING_32')):
                    self.messageId = StringID.stringID(unwrapper.unwrapString(element))
                elif code in (AMQPType.amqpType.getValueByKey('BINARY_8'),AMQPType.amqpType.getValueByKey('BINARY_32')):
                    self.messageId = BinaryID.binaryID(unwrapper.unwrapBinary(element))
                elif code in (AMQPType.amqpType.getValueByKey('UUID')):
                    self.messageId = BinaryID.binaryID(unwrapper.unwrapUuid(element))
                else:
                    print('Expected type MessageID received ' + str(element.getCode()))
        if len(list.getList()) > 1:
            element = list.getList()[1]
            if element is not None:
                self.userId = unwrapper.unwrapBinary(element)
        if len(list.getList()) > 2:
            element = list.getList()[2]
            if element is not None:
                self.to = unwrapper.unwrapString(element)
        if len(list.getList()) > 3:
            element = list.getList()[3]
            if element is not None:
                self.subject = unwrapper.unwrapString(element)
        if len(list.getList()) > 4:
            element = list.getList()[4]
            if element is not None:
                self.replyTo = unwrapper.unwrapString(element)
        if len(list.getList()) > 5:
            element = list.getList()[5]
            if element is not None:
                self.correlationId = unwrapper.unwrapBinary(element)
        if len(list.getList()) > 6:
            element = list.getList()[6]
            if element is not None:
                self.contentType = unwrapper.unwrapString(element)
        if len(list.getList()) > 7:
            element = list.getList()[7]
            if element is not None:
                self.contentEncoding = unwrapper.unwrapString(element)
        if len(list.getList()) > 8:
            element = list.getList()[8]
            if element is not None:
                self.absoluteExpiryTime = unwrapper.unwrapTimastamp(element)
        if len(list.getList()) > 9:
            element = list.getList()[9]
            if element is not None:
                self.creationTime = unwrapper.unwrapTimastamp(element)
        if len(list.getList()) > 10:
            element = list.getList()[10]
            if element is not None:
                self.groupId = unwrapper.unwrapString(element)
        if len(list.getList()) > 11:
            element = list.getList()[11]
            if element is not None:
                self.groupSequence = unwrapper.unwrapUInt(element)
        if len(list.getList()) > 12:
            element = list.getList()[12]
            if element is not None:
                self.replyToGroupId = unwrapper.unwrapString(element)

    def getCode(self):
        return SectionCode.sectionCode.getValueByKey('PROPERTIES')

    def toString(self):
        return 'AMQPProperties [messageId=' + self.messageId + ', userId=' + self.userId + ', to=' + self.to + ', subject=' + self.subject + ', replyTo=' + self.replyTo + ', correlationId=' + self.correlationId + ', contentType=' + self.contentType + ', contentEncoding=' + self.contentEncoding + ', absoluteExpiryTime=' + self.absoluteExpiryTime + ', creationTime=' + self.creationTime + ', groupId=' + self.groupId + ', groupSequence=' + self.groupSequence + ', replyToGroupId=' + self.replyToGroupId + ']'

    def getMessageId(self):
        return self.messageId

    def setMessageId(self, messageId):
        self.messageId = messageId

    def getUserId(self):
        return self.userId

    def setUserId(self, userId):
        self.userId = userId

    def getTo(self):
        return self.to

    def setTo(self, to):
        self.to = to

    def getSubject(self):
        return self.subject

    def setSubject(self, subject):
        self.subject = subject

    def getReplyTo(self):
        return self.replyTo

    def setReplyTo(self, replyTo):
        self.replyTo = replyTo

    def getCorrelationId(self):
        return self.correlationId

    def setCorrelationId(self, correlationId):
        self.correlationId = correlationId

    def getContentType(self):
        return self.contentType

    def setContentType(self, contentType):
        self.contentType = contentType

    def getContentEncoding(self):
        return self.contentEncoding

    def setContentEncoding(self, contentEncoding):
        self.contentEncoding = contentEncoding

    def getAbsoluteExpiryTime(self):
        return self.absoluteExpiryTime

    def setAbsoluteExpiryTime(self, absoluteExpiryTime):
        self.absoluteExpiryTime = absoluteExpiryTime

    def getCreationTime(self):
        return self.creationTime

    def setCreationTime(self, creationTime):
        self.creationTime = creationTime

    def getGroupId(self):
        return self.groupId

    def setGroupId(self, groupId):
        self.groupId = groupId

    def getGroupSequence(self):
        return self.groupSequence

    def setGroupSequence(self, groupSequence):
        self.groupSequence = groupSequence

    def getReplyToGroupId(self):
        return self.replyToGroupId

    def setReplyToGroupId(self, replyToGroupId):
        self.replyToGroupId = replyToGroupId

File: IoTbrokerClient/src/iot-protocols/amqp/classes/section/amqpproperties.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpproperties.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/wrappers/messageID/amqpbinaryid.h"
#include "iot-protocols/amqp/classes/wrappers/messageID/amqplongid.h"
#include "iot-protocols/amqp/classes/wrappers/messageID/amqpmessageid.h"
#include "iot-protocols/amqp/classes/wrappers/messageID/amqpstringid.h"
#include "iot-protocols/amqp/classes/wrappers/messageID/amqpuuid.h"

AMQPProperties::AMQPProperties()
{

}

TLVAMQP *AMQPProperties::getValue()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->messageId != NULL) {
        JsonVariant *object = NULL;
        if (this->messageId->getBinary() != NULL) {
            object = AMQPSimpleType::binaryToVariant(this->messageId->getBinary());
        } else if (this->messageId->getLongValue() != NULL) {
            object = AMQPSimpleType::longToVariant(this->messageId->getLongValue());
        } else if (this->messageId->getString() != NULL) {
            object = AMQPSimpleType::stringToVariant(this->messageId->getString());
        } else if (this->messageId->getUUID() != NULL) {
            object = AMQPSimpleType::uuidToVariant(this->messageId->getUUID());
        }
        list->addElementWithIndex(0, AMQPWrapper::wrapObject(object));
    }
    if (this->userId != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapBinary(this->userId));
    }
    if (this->to != NULL) {
        list->addElementWithIndex(2, AMQPWrapper::wrapString(this->to));
    }
    if (this->subject != NULL) {
        list->addElementWithIndex(3, AMQPWrapper::wrapString(this->subject));
    }
    if (this->replyTo != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapString(this->replyTo));
    }
    if (this->correlationID != NULL) {
        list->addElementWithIndex(5, AMQPWrapper::wrapBinary(this->correlationID));
    }
    if (this->contentType != NULL) {
        list->addElementWithIndex(6, AMQPWrapper::wrapString(this->contentType));
    }
    if (this->contentEncoding != NULL) {
        list->addElementWithIndex(7, AMQPWrapper::wrapString(this->contentEncoding));
    }
    if (this->absoluteExpiryTime != NULL) {
        list->addElementWithIndex(8, AMQPWrapper::wrapTimestamp(this->absoluteExpiryTime));
    }
    if (this->creationTime != NULL) {
        list->addElementWithIndex(9, AMQPWrapper::wrapTimestamp(this->creationTime));
    }
    if (this->groupId != NULL) {
        list->addElementWithIndex(10, AMQPWrapper::wrapString(this->groupId));
    }
    if (this->groupSequence != NULL) {
        list->addElementWithIndex(11, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->groupSequence)));
    }
    if (this->replyToGroupId != NULL) {
        list->addElementWithIndex(12, AMQPWrapper::wrapString(this->replyToGroupId));
    }

    unsigned char * data = NULL;
    data = malloc(sizeof(char));
    data[0] = (char) 0x73;
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);
    return list;
}

AMQPSectionCode *AMQPProperties::getCode()
{
    return new AMQPSectionCode(AMQP_PROPERTIES_SECTION_CODE);
}

void AMQPProperties::fill(TLVAMQP *value)
{
    AMQPTLVList *list = new AMQPTLVList();

    if (list->getList().size() > 0) {
        TLVAMQP *element = list->getList()[0];
        if (!element->isNull()) {
            switch (element->getTypeValue()) {
                case AMQP_ULONG0_TYPE:
                case AMQP_SMALL_ULONG_TYPE:
                case AMQP_ULONG_TYPE:
                {
                    long long value = AMQPUnwrapper::unwrapLong(element);
                    JsonVariant variant;
                    variant.set(value);
                    this->messageId = new AMQPLongID(variant);
                } break;
                case AMQP_STRING8_TYPE:
                case AMQP_STRING32_TYPE:
                {
                    this->messageId = new AMQPStringID(AMQPUnwrapper::unwrapString(element));
                } break;
                case AMQP_BINARY8_TYPE:
                case AMQP_BINARY32_TYPE:
                {
                    this->messageId = new AMQPBinaryID(AMQPUnwrapper::unwrapBinary(element));
                } break;
                case AMQP_UUID_TYPE:
                {
                    this->messageId = new AMQPUUID(AMQPUnwrapper::unwrapUUID(element));
                } break;
                default: break;
            }
        }
    }
    if (list->getList().size() > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            this->userId = AMQPUnwrapper::unwrapBinary(element);
        }
    }
    if (list->getList().size() > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            this->to = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 3) {
        TLVAMQP *element = list->getList()[3];
        if (!element->isNull()) {
            this->subject = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 4) {
        TLVAMQP *element = list->getList()[4];
        if (!element->isNull()) {
            this->replyTo = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 5) {
        TLVAMQP *element = list->getList()[5];
        if (!element->isNull()) {
            this->correlationID = AMQPUnwrapper::unwrapBinary(element);
        }
    }
    if (list->getList().size() > 6) {
        TLVAMQP *element = list->getList()[6];
        if (!element->isNull()) {
            this->contentType = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 7) {
        TLVAMQP *element = list->getList()[7];
        if (!element->isNull()) {
            this->contentEncoding = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 8) {
        TLVAMQP *element = list->getList()[8];
        if (!element->isNull()) {
            this->absoluteExpiryTime = AMQPUnwrapper::unwrapTimestamp(element);
        }
    }
    if (list->getList().size() > 9) {
        TLVAMQP *element = list->getList()[9];
        if (!element->isNull()) {
            this->creationTime = AMQPUnwrapper::unwrapTimestamp(element);
        }
    }
    if (list->getList().size() > 10) {
        TLVAMQP *element = list->getList()[10];
        if (!element->isNull()) {
            this->groupId = AMQPUnwrapper::unwrapUInt(element);
        }
    }
    if (list->getList().size() > 11) {
        TLVAMQP *element = list->getList()[11];
        if (!element->isNull()) {
            this->groupSequence = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (list->getList().size() > 12) {
        TLVAMQP *element = list->getList()[12];
        if (!element->isNull()) {
            this->replyToGroupId = AMQPUnwrapper::unwrapString(element);
        }
    }
}

AMQPMessageID *AMQPProperties::getMessageId() const
{
    return messageId;
}

void AMQPProperties::setMessageId(AMQPMessageID *value)
{
    messageId = value;
}

unsigned char * AMQPProperties::getUserId() const
{
    return userId;
}

void AMQPProperties::setUserId(const unsigned char &value)
{
    userId = value;
}

String AMQPProperties::getTo() const
{
    return to;
}

void AMQPProperties::setTo(const String &value)
{
    to = value;
}

String AMQPProperties::getSubject() const
{
    return subject;
}

void AMQPProperties::setSubject(const String &value)
{
    subject = value;
}

String AMQPProperties::getReplyTo() const
{
    return replyTo;
}

void AMQPProperties::setReplyTo(const String &value)
{
    replyTo = value;
}

unsigned char * AMQPProperties::getCorrelationID() const
{
    return correlationID;
}

void AMQPProperties::setCorrelationID(const unsigned char &value)
{
    correlationID = value;
}

String AMQPProperties::getContentType() const
{
    return contentType;
}

void AMQPProperties::setContentType(const String &value)
{
    contentType = value;
}

String AMQPProperties::getContentEncoding() const
{
    return contentEncoding;
}

void AMQPProperties::setContentEncoding(const String &value)
{
    contentEncoding = value;
}

time_t AMQPProperties::getAbsoluteExpiryTime() const
{
    return absoluteExpiryTime;
}

void AMQPProperties::setAbsoluteExpiryTime(const time_t &value)
{
    absoluteExpiryTime = value;
}

time_t AMQPProperties::getCreationTime() const
{
    return creationTime;
}

void AMQPProperties::setCreationTime(const time_t &value)
{
    creationTime = value;
}

String AMQPProperties::getGroupId() const
{
    return groupId;
}

void AMQPProperties::setGroupId(const String &value)
{
    groupId = value;
}

JsonVariant *AMQPProperties::getGroupSequence() const
{
    return groupSequence;
}

void AMQPProperties::setGroupSequence(JsonVariant *value)
{
    groupSequence = value;
}

String AMQPProperties::getReplyToGroupId() const
{
    return replyToGroupId;
}

void AMQPProperties::setReplyToGroupId(const String &value)
{
    replyToGroupId = value;
}
